"Как я использовал(а) бутстрэп ? "

Из лекции я поняла, что bootstrap может являться одним из эффективных механизмов для того, 
чтобы оценивать выборочное распределение стат. данных/параметров внутри нашей модели.

Удобно,что нам совсем не обязательно иметь выполненную предпосылку о нормальности распределения.

In [ ]:
#Пробую использовать уже оптимизированную и удобную библиотеку для бутстрэпа:
#!pip3 install arch

In [8]:
conda install -c conda-forge arch-py

done
Solving environment: done

## Package Plan ##

  environment location: /Users/annasidorova/opt/anaconda3

  added / updated specs:
    - arch-py


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-4.14.0               |   py38h50d1736_0        1012 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        1012 KB

The following packages will be UPDATED:

  conda                               4.12.0-py38h50d1736_0 --> 4.14.0-py38h50d1736_0



conda-4.14.0         | 1012 KB   | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done

Note: you may need to restart the kernel to use updated packages.


In [ ]:
#pip install --upgrade numpy

In [11]:
import numpy as np
np.bool = np.bool_

In [13]:
conda install pandas

Solving environment: done

# All requested packages already installed.

Retrieving notices: ...working... done

Note: you may need to restart the kernel to use updated packages.


In [19]:
#pip install numpy==1.19

     |████████████████████████████████| 15.3 MB 13.4 MB/s eta 0:00:01   |█████▎                          | 2.5 MB 1.6 MB/s eta 0:00:09     |███████████████████▋            | 9.4 MB 518 kB/s eta 0:00:12
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.4
    Uninstalling numpy-1.24.4:
      Successfully uninstalled numpy-1.24.4
Note: you may need to restart the kernel to use updated packages.


In [20]:
conda install seaborn

Solving environment: done

# All requested packages already installed.

Retrieving notices: ...working... done

Note: you may need to restart the kernel to use updated packages.


In [40]:
conda install -c conda-forge bootstrap

done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.14.0
  latest version: 23.10.0

Please update conda by running

    $ conda update -n base -c conda-forge conda



## Package Plan ##

  environment location: /Users/annasidorova/opt/anaconda3

  added / updated specs:
    - bootstrap


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bootstrap-4.4.1            |                0         590 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         590 KB

The following NEW packages will be INSTALLED:

  bootstrap          conda-forge/noarch::bootstrap-4.4.1-0



bootstrap-4.4.1      | 590 KB    | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Retrieving notices: ...working..

In [24]:
#Модифицирую пример из лекции. Инициилизируем генератор случайных чисел:
random = np.random.default_rng(1000)
#Зерно генератора - задали равным 1000.

С помощью созданного генератора "random" придумаем псевдослучайную выборку из нормального распределения с мат ожиданием равным 4 и стандартным отклонением равным 3.

Возьмем n=10 в качестве кол-ва наблюдений.

In [25]:
x = random.normal(loc = 4, scale = 3, size = 10)

In [26]:
#Как выглядит наша случайная выборка x?
x

array([3.03600938, 2.54301557, 9.04017439, 9.91158065, 4.46555098,
       0.22670366, 4.60065525, 4.12615898, 1.3659506 , 8.58822369])

In [27]:
np.mean(x)

4.790402313117086

Видим, что выборочные характеристики отличаются от истинных: 
реальное мат ожидание каждого икса (равное 4) vs выборочное среднее (равное 4.79).

Допустим, что мы не знаем истинное значение мат ожидания. Хотим получить точечную оценку.

In [29]:
#Естественная точечная оценка для неизвестного мат ожидания:
w_hat = np.mean(x)
w_hat

4.790402313117086

In [31]:
# А что делать для построения Д.И. для неизвестного мат ожидания? (если не знаем формулу) 
# Превращаем нашу точечную оценку в интервальную:

# Первая фейковая выборка - из 10 наблюдений берем 10 с возможностью повтора:
x_star1 = random.choice(x, size = len(x))
x_star1


array([1.3659506 , 1.3659506 , 4.60065525, 9.04017439, 4.12615898,
       2.54301557, 4.60065525, 8.58822369, 2.54301557, 4.60065525])

In [32]:
# Первая симулированная точечная оценка:
np.mean(x_star1)

4.337445511226076

И ведь таких бустрэп-выборок можно создать много!

In [34]:
number_of_boot = 1800 # Хотим получить 1800 выборок из нашей исходной выборки
# Поместим эти копии в список:
w_hat_star = [np.mean(random.choice(x, size = len(x))) for i in range(number_of_boot)]
# 1800 точечных оценок для мат ожидания!

In [36]:
# Посмотрим срез полученных данных по точечным оценкам:
w_hat_star[1:5]

[5.098107794511509, 5.848474728749716, 5.384304660983801, 2.4895421813417906]

In [37]:
# Покажем, как в нашем случае будет работать наивный бутстрэп. 
## для дефолтного двухстороннего Д.И. для неизвестного мат ожидания:

# Взяли квантиль от нашей выборки порядка 0.025 левый и 0.975 правый, 
## чтобы получить доверительный интервал для мат ожидания:
[np.quantile(w_hat_star,0.025), np.quantile(w_hat_star, 0.975)]

[2.8329849379797913, 6.824890319712152]

Настоящее мат ожидание равнялось 4, и мы накрыли его в нашем доверительном интервале!

Вспомним, что после каждой выемки элемента мы его обратно возвращаем.
Таким образом получаем выводы:
   
   1.(Наивный) бутстрэп - средство для создания многочисленной популяции элементов 
    (которая повторяет в себе характеристики изначальной выборки).
   2. Чем больше итераций мы выполним, тем точнее получаем оценку нашего доверительного интервала.

In [ ]:
#Пример Basic IID Bootstrap , похожий на задачку выше:

###from arch.bootstrap import IIDBootstrap, IndependentSamplesBootstrap
#bstr_x = IIDBootstrap(x, seed = 100000) # задали ядро равным 100000
#bstr_x.conf_int(np.mean, method = 'basic', reps = 20000, size = 0.95)